packages/@react-stately/utils/src/useControlledState.ts

```diff
@@ -47,15 +47,10 @@ export function useControlledState<T>(
       // this will call our useState setState with a function as well which invokes myFunc and calls onChange with the value from myFunc
       // if we're in an uncontrolled state, then we also return the value of myFunc which to setState looks as though it was just called with myFunc from the beginning
       // otherwise we just return the controlled value, which won't cause a rerender because React knows to bail out when the value is the same
-      let updateFunction = (oldValue, ...functionArgs) => {
-        let interceptedValue = value(isControlled ? stateRef.current : oldValue, ...functionArgs);
-        onChangeCaller(interceptedValue, ...args);
-        if (!isControlled) {
-          return interceptedValue;
-        }
-        return oldValue;
-      };
-      setStateValue(updateFunction);
+      let oldValue = stateRef.current;
+      let interceptedValue = value(oldValue);
+      setStateValue(!isControlled ? interceptedValue : oldValue);
+      onChangeCaller(interceptedValue, ...args);
```

Comment:
What was the reason for changing this? Isn't the point of the old way that the update function isn't called until the actual state updates in React and to handle batching?

Reply:
it was complaining about me changing state in a component during its render cycle (usually something i’d expect in react 17) however, the message was the same as one that Niklas ran into in one of his PR’s to address strictmode

Reply:
it's also an issue here https://github.com/adobe/react-spectrum/issues/1223

Reply:
I think this will change the behavior of useControlledState in subtle ways. Can we tackle this as a separate PR maybe? If it's only causing a warning, and that warning already exists, then maybe it can be treated as a separate issue?

Reply:
Agreed, it makes me nervous as well, but it hasn't changed anything that I've been able to find, and I've been staring at it for a few months